### PR TITLE
Added target go-vulndb

### PR DESCRIPTION
**Pull request by rahul2393**
[Issue](https://github.com/aquasecurity/vuln-list-update/issues/82)

### CODE REVIEW

File: .github/workflows/update.yml [hunk 1]
```diff
@@ -124,3 +124,7 @@ jobs:
     - if: always()
       name: Red Hat Security Data API 2019-2021
       run: ./vuln-list-update -target redhat -years 2019,2020,2021
+
+    - if: always()
+      name: GoVulnDB
```
Comment (knqyf263) [inline .github/workflows/update.yml]: Please put it before Red Hat.

Comment (knqyf263) [reply .github/workflows/update.yml]: nit
```suggestion
      name: Go Vulnerability Database
```

File: govulndb/govulndb.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+const (
+	baseURL     = "https://storage.googleapis.com/go-vulndb"
+	goVulnDBDir = "go"
+	concurrency = 5
+	wait        = 0
+	retry       = 5
+)
+
+func Update() error {
+	log.Println("Fetching GoVulnDB data...")
```
Comment (knqyf263) [inline govulndb/govulndb.go:24]:
```suggestion
	log.Println("Fetching Go Vulnerability Database...")
```

File: govulndb/govulndb.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	b, err := utils.FetchURL(baseURL+"/index.json", "", retry)
```
Comment (knqyf263) [inline govulndb/govulndb.go:26]: Ideally, we should use `path.Join`.
https://stackoverflow.com/questions/34668012/combine-url-paths-with-path-join/34668130

File: govulndb/govulndb.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	if err != nil {
+		return err
```
Comment (knqyf263) [inline govulndb/govulndb.go:28]: Please wrap the error by `xerrors.Errorf`

File: govulndb/govulndb.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	}
+	var vulnerablePackages map[string]string
+	if err := json.Unmarshal(b, &vulnerablePackages); err != nil {
+		return xerrors.Errorf("failed to decode goVulnDB index.json response: %w", err)
```
Comment (knqyf263) [inline govulndb/govulndb.go:32]: nit: 
```suggestion
		return xerrors.Errorf("failed to decode go-vulndb index.json response: %w", err)
```

File: govulndb/govulndb.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	}
+	urls := make([]string, len(vulnerablePackages))
+	i := 0
+	for packageName := range vulnerablePackages {
+		url := fmt.Sprintf("%s/%s.json", baseURL, packageName)
+		urls[i] = url
+		i++
+	}
```
Comment (knqyf263) [inline govulndb/govulndb.go:40]: `vulnerablePackages` is small enough. We can simply use `append` for simplicity .in this case.

```suggestion
	var urls []string
	for packageName := range vulnerablePackages {
		url := fmt.Sprintf("%s/%s.json", baseURL, packageName)
		urls = append(urls, url)
	}
```

File: govulndb/govulndb.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	responses, err := utils.FetchConcurrently(urls, concurrency, wait, retry)
+	if err != nil {
+		return xerrors.Errorf("failed to fetch concurrently: %w", err)
+	}
+	log.Println("Saving GoVulnDB data...")
```
Comment (knqyf263) [inline govulndb/govulndb.go:45]: nit
```suggestion
	log.Println("Saving Go Vulnerability Database...")
```

File: govulndb/govulndb.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	bar := pb.StartNew(len(responses))
+	for _, res := range responses {
+		var entries []Entry
+		if err := json.Unmarshal(res, &entries); err != nil {
+			return xerrors.Errorf("failed to decode goVulnDB response: %w", err)
```
Comment (knqyf263) [inline govulndb/govulndb.go:50]:
```suggestion
			return xerrors.Errorf("failed to decode go-vulndb response: %w", err)
```

File: govulndb/govulndb.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+		}
+		if err := save(entries); err != nil {
+			return err
```
Comment (knqyf263) [inline govulndb/govulndb.go:53]: wrap the error

File: govulndb/govulndb.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+		bar.Increment()
+	}
+	bar.Finish()
+	return nil
+}
+
+func save(entries []Entry) error {
+	for _, entry := range entries {
+		cveID := entry.ID
+		if err := utils.SaveCVEPerYear(fmt.Sprintf("%s/%s", goVulnDBDir, entry.Package.Name), cveID, entry); err != nil {
```
Comment (knqyf263) [inline govulndb/govulndb.go:64]: We should not use `SaveCVEPerYear`. Store vulnerabilities grouped by packages.

File: govulndb/govulndb.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+			return xerrors.Errorf("failed to save NVD CVE detail: %w", err)
```
Comment (knqyf263) [inline govulndb/govulndb.go:65]: Not NVD

File: main.go [hunk 12]
```diff
@@ -171,6 +172,11 @@ func run() error {
 			return xerrors.Errorf("error in CWE update: %w", err)
 		}
 		commitMsg = "CWE Advisories"
+	case "go-vulndb":
+		if err := govulndb.Update(); err != nil {
+			return xerrors.Errorf("error in GoVulnDb update: %w", err)
+		}
+		commitMsg = "GO Vulndb"
```
Comment (knqyf263) [inline main.go:179]: nit
```suggestion
		commitMsg = "Go Vulnerability Database"
```

File: main.go [hunk 13]
```diff
@@ -171,6 +172,11 @@ func run() error {
 			return xerrors.Errorf("error in CWE update: %w", err)
 		}
 		commitMsg = "CWE Advisories"
+	case "go-vulndb":
+		if err := govulndb.Update(); err != nil {
+			return xerrors.Errorf("error in GoVulnDb update: %w", err)
```
Comment (knqyf263) [inline main.go:177]: nit
```suggestion
			return xerrors.Errorf("error in go-vulndb update: %w", err)
```

File: govulndb/types.go [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+	GOOS    []string `json:",omitempty"`
+	GOARCH  []string `json:",omitempty"`
+	URL     string
+}
+
+type Entry struct {
+	ID         string
+	Published  time.Time
+	Modified   time.Time
+	Withdrawn  *time.Time `json:",omitempty"`
```
Comment (knqyf263) [inline govulndb/types.go:39]: Why is this a pointer?

**Review by knqyf263 — CHANGES_REQUESTED** [review]
Thanks! I left some comments. Also, write tests, please.

File: govulndb/types.go (hunk 14 above)
Comment (rahul2393) [reply govulndb/types.go:39]: Source: https://github.com/golang/vulndb/blob/e0c00fae09e687ec6febda47ae3bc7552fc7b988/osv/json.go#L125

File: govulndb/govulndb.go [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+	"path/filepath"
+
+	"github.com/cheggaaa/pb/v3"
+	"golang.org/x/xerrors"
+
+	"github.com/aquasecurity/vuln-list-update/utils"
+)
+
+const (
+	vulndbURL = "https://storage.googleapis.com/go-vulndb"
```
Comment (patryk4815) [inline govulndb/govulndb.go:18]: why not using https://osv-vulnerabilities.storage.googleapis.com/Go/all.zip ?
https://github.com/google/osv/blob/master/README.md#current-data-sources

**Comment by knqyf263** [comment]
In favor of https://github.com/aquasecurity/vuln-list-update/pull/120

**Pull request closed without merging**